File: arena/internal/models/arena.go
package models

import "time"

type ArenaRoomStatus string

const (
	StatusWaiting  ArenaRoomStatus = "WAITING"
	StatusLobby    ArenaRoomStatus = "LOBBY"
	StatusPlaying  ArenaRoomStatus = "PLAYING"
	StatusFinished ArenaRoomStatus = "FINISHED"
)

type ArenaPlayerStatus string

const (
	PlayerCoding    ArenaPlayerStatus = "CODING"
	PlayerSubmitted ArenaPlayerStatus = "SUBMITTED"
)

type ArenaPlayer struct {
	UserID      string            `json:"userId"`
	Username    string            `json:"username"`
	AvatarURL   string            `json:"avatarUrl,omitempty"`
	IsCreator   bool              `json:"isCreator"`
	Score       int               `json:"score"`
	TestsPassed int               `json:"testsPassed"`
	TotalTests  int               `json:"totalTests"`
	SubmittedAt *time.Time        `json:"submittedAt,omitempty"`
	Status      ArenaPlayerStatus `json:"status"`
	IsOffline       bool              `json:"isOffline"`
	JoinedAt        time.Time         `json:"joinedAt"`
	SubmissionOrder int               `json:"submissionOrder"`
	TimeTaken       int               `json:"timeTaken,omitempty"`
}

type ArenaRoom struct {
	RoomID      string                 `json:"roomId"`
	Status      ArenaRoomStatus        `json:"status"`
	Topic       string                 `json:"topic,omitempty"`
	ProblemID   string                 `json:"problemId,omitempty"`
	ProblemSlug string                 `json:"problemSlug,omitempty"`
	Difficulty    string                 `json:"difficulty,omitempty"`
	Language      string                 `json:"language,omitempty"`
	MatchDuration int                    `json:"matchDuration,omitempty"`
	Players       map[string]ArenaPlayer `json:"players"`
	CreatedAt   time.Time              `json:"createdAt"`
	StartTime   *int64                 `json:"startTime,omitempty"`
	EndTime     *int64                 `json:"endTime,omitempty"`
	WinnerID    string                 `json:"winnerId,omitempty"`
}

type ArenaWSMessage struct {
	Type    string      `json:"type"`
	Payload interface{} `json:"payload"`
}

File: web/src/types/arena.ts
export type ArenaRoomStatus = "WAITING" | "PLAYING" | "FINISHED" | "LOBBY";

/**
 * Transient Player State (Redis/WebSocket)
 * Used for real-time updates in the store.
 */
export interface ArenaPlayer {
  userId: string;
  username: string;
  avatarUrl?: string;
  isCreator: boolean;
  score: number;
  testsPassed: number;
  totalTests: number;
  joinedAt: string;
  status: "CODING" | "SUBMITTED";
  isOffline: boolean;
  rank?: number; // Real-time ranking during match
  submissionOrder?: number; // Order of successful submission
  timeTaken?: number;
}

/**
 * Room Metadata (MongoDB Client View)
 */
export interface ArenaRoom {
  roomId: string;
  status: ArenaRoomStatus;
  topic?: string;
  problemId?: string;
  problemSlug?: string;
  difficulty?: string;
  language?: string;
  matchDuration?: number;
  players: Record<string, ArenaPlayer>;
  createdAt: string;
  startTime?: number;
  endTime?: number;
  matchId?: string;
}

export interface ArenaWSMessage {
  type:
    | "PLAYER_JOINED"
    | "PLAYER_LEFT"
    | "PLAYER_READY"
    | "START_MATCH"
    | "MATCH_START"
    | "MATCH_STARTED"
    | "PROGRESS_UPDATE"
    | "MATCH_SUBMITTED"
    | "MATCH_OVER"
    | "MATCH_ENDED"
    | "PROBLEM_CHANGED"
    | "ERROR"
    | "CHAT"
    | "LEAVE_ROOM"
    | "PLAYER_REMOVED"
    | "LEADERBOARD_UPDATE"
    | "HOST_TRANSFERRED"
    | "UPDATE_MATCH_DURATION"
    | "MATCH_DURATION_CHANGED"
    | "KICK_PLAYER"
    | "YOU_ARE_KICKED"
    | "PLAYER_KICKED"
    | "ABORT_MATCH";
  payload: any;
}

/**
 * Persistent Player Result (MongoDB)
 * Used for historical match records and final rankings.
 */
export interface ArenaPlayerResult {
  userId: string;
  username: string;
  avatarUrl?: string;
  finalRank?: number;
  score: number;
  testsPassed: number;
  totalTests: number;
  verdict: "ACCEPTED" | "WRONG_ANSWER" | "TIME_LIMIT_EXCEEDED" | "MEMORY_LIMIT_EXCEEDED" | "RUNTIME_ERROR" | "COMPILATION_ERROR" | "SKIPPED";
  submittedAt?: string;
  submissionOrder?: number;
  sourceCode?: string;
  languageId?: string;
  timeTaken?: number;
}

/**
 * Match Record (MongoDB)
 */
export interface ArenaMatch {
  id: string;
  roomId: string;
  hostId: string;
  problemId: string;
  problemSlug: string;
  difficulty: "Easy" | "Medium" | "Hard";
  language: string;
  status: "FINISHED" | "CANCELLED";
  players: ArenaPlayerResult[];
  startedAt: string;
  endedAt?: string;
}
